File: articles/002_7-Messages-CSharp/ObjectClasses/Checkout.cs

using Messages.Taxonomy;

namespace Messages.ObjectClasses
{

    public class Checkout : IMessageHandler
    {

        private CartTotalsUpdated _latestTotals = new(0, 0);

        public IEnumerable<Message>? On(Message message) => message switch
        {
            CartTotalsUpdated cartTotalsUpdated => OnCartTotalsUpdated(cartTotalsUpdated),
            AtCheckout _ => OnAtCheckout(),
            _ => null
        };

        private IEnumerable<Message>? OnAtCheckout()
        {
            Console.WriteLine($"{_latestTotals.ItemCount} items, total cost: {_latestTotals.TotalCost}");
            return null;
        }

        private IEnumerable<Message>? OnCartTotalsUpdated(CartTotalsUpdated cartTotalsUpdated)
        {
            _latestTotals = cartTotalsUpdated with {};
            return null;
        }
    }
}

File: articles/003-OutsideAndInside/factories/Checkout.js
import { AtCheckout, CartTotalsUpdated } from "../taxonomy.js";

export function Checkout() {

    let latestTotals;

    return message => {

        switch (true) {
            case message instanceof CartTotalsUpdated:
                latestTotals = message;
                break;
            case message instanceof AtCheckout:
                logCartTotals(latestTotals);
                break;
        }

    };

}

function logCartTotals({ items, total }) {

    console.log(items, "items, total cost:", total);

}
